Introduce a file type template

tenderlove opened the pull request:
Every template that specifies a "virtual path" loses the template source
when the template gets compiled:

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/template.rb#L275

The "refresh" method seems to think that the source code for a template
can be recovered if there is a virtual path:

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/template.rb#L171-L188

Every call site that allocates a template object *and* provides a
"virtual path" reads the template contents from the filesystem:

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/template/resolver.rb#L229-L231

Templates that are inline or literals don't provide a "virtual path":

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/renderer/template_renderer.rb#L34

This commit introduces a `FileTemplate` type that subclasses `Template`.
The `FileTemplate` keeps a reference to the filename, and reads the
source from the filesystem.  This effectively makes the template source
immutable.

Other classes depended on the source to be mutated while being compiled,
so this commit also introduces a temporary way to pass the mutated
source to the ERB (or whatever) compiler.  See `LegacyTemplate`.

I think we should consider it an error to provide a virtual path on a
non file type template an non-file templates can't recover their source.
Here is an example:

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/testing/resolvers.rb#L53

This provides a "virtual path" so the source code (a string literal) is
thrown away after compilation.  Clearly we can't recover that string, so
I think this should be an error.

tenderlove committed:
use the source returned from encode!

tenderlove committed:
Introduce a file type template, deprecate `Template#refresh`

Every template that specifies a "virtual path" loses the template source
when the template gets compiled:

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/template.rb#L275

The "refresh" method seems to think that the source code for a template
can be recovered if there is a virtual path:

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/template.rb#L171-L188

Every call site that allocates a template object *and* provides a
"virtual path" reads the template contents from the filesystem:

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/template/resolver.rb#L229-L231

Templates that are inline or literals don't provide a "virtual path":

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/renderer/template_renderer.rb#L34

This commit introduces a `FileTemplate` type that subclasses `Template`.
The `FileTemplate` keeps a reference to the filename, and reads the
source from the filesystem.  This effectively makes the template source
immutable.

Other classes depended on the source to be mutated while being compiled,
so this commit also introduces a temporary way to pass the mutated
source to the ERB (or whatever) compiler.  See `LegacyTemplate`.

I think we should consider it an error to provide a virtual path on a
non file type template an non-file templates can't recover their source.
Here is an example:

  https://github.com/rails/rails/blob/eda0f574f129fcd5ad1fc58b55cb6d1db71ea95c/actionview/lib/action_view/testing/resolvers.rb#L53

This provides a "virtual path" so the source code (a string literal) is
thrown away after compilation.  Clearly we can't recover that string, so
I think this should be an error.

rafaelfranca commented on actionview/lib/action_view/template.rb:
```diff
@@ -270,13 +270,19 @@ def compile!(view)
         end
       end
 
+      class LegacyTemplate < DelegateClass(Template) # :nodoc:
```
What is the plan to this class? We will always have a LegacyTemplate or we plan to remove it some day?

tenderlove replied:
I think we can remove it eventually.  I want to change [this call](https://github.com/rails/rails/blob/2169bd3d2a9d2f331a5dd6e41d9d638e0da6117c/actionview/lib/action_view/template.rb#L307) to pass in the view and the source.  But for backwards compatibility we need to pass an object that has a reference to the mutated source.

I'll follow up with another commit that deprecates arity one blocks

tenderlove commented:
I changed this to not deprecate `refresh`.  I think we might be able to deprecate and remove it in the future, but right now it's not clear.  In test, we have a custom resolver that instantiates template objects that have a view path, but aren't associated with a file:

  https://github.com/rails/rails/blob/2169bd3d2a9d2f331a5dd6e41d9d638e0da6117c/actionview/lib/action_view/testing/resolvers.rb#L26-L47

After this PR, the common case will be to instantiate `FileTemplate` objects, so I just implemented a `refresh` on `FileTemplate` that returns itself (since it can get the source).

rafaelfranca commented:
> I changed this to not deprecate `refresh`. I think we might be able to deprecate and remove it in the future, but right now it's not clear. In test, we have a custom resolver that instantiates template objects that have a view path

Ah right. This is something we support and José's book talk about it in the "Retrieving View Templates from Custom Stores" chapter https://pragprog.com/book/jvrails2/crafting-rails-4-applications

tenderlove committed:
Pass source to template handler and deprecate old style handler

This commit passes the mutated source to the template handler as a
parameter and deprecates the old handlers.  Old handlers required that
templates contain a reference to mutated source code, but we would like
to make template objects "read only".  This change lets the template
remain "read only" while still giving template handlers access to the
source code after mutations.

morgoth commented:
While upgrading to Rails 6.0 I'm getting warning:
```
DEPRECATION WARNING: Single arity template handlers are deprecated.  Template handlers must
now accept two parameters, the view object and the source for the view object.
Change:
  >> Class#call(template)
To:
  >> Class#call(template, source)
 (called from <top (required)> at /home/wojtek/Projects/some_project/config/environment.rb:5)
```
which is not really clear where to find it. I suspect some gem, as I don't mess with templates in the app.
Do you have a clue how to track it?

Sun-Mountain commented:
> While upgrading to Rails 6.0 I'm getting warning:
> 
> ```
> DEPRECATION WARNING: Single arity template handlers are deprecated.  Template handlers must
> now accept two parameters, the view object and the source for the view object.
> Change:
>   >> Class#call(template)
> To:
>   >> Class#call(template, source)
>  (called from <top (required)> at /home/wojtek/Projects/some_project/config/environment.rb:5)
> ```
> 
> which is not really clear where to find it. I suspect some gem, as I don't mess with templates in the app.
> Do you have a clue how to track it?

I have the same question. I am now updating our rails app and getting the same message but cannot find anywhere this might affect our app. (We also do not use templates outside of the devise gem.)

eugeneius commented:
Setting `config.active_support.deprecation = :raise` in `config/environments/development.rb` and trying to run the application will give you a backtrace that includes the line that's causing the problem. It's most likely due to an older version of the jbuilder or coffee-script gems.

I think the code that tries to extract the culprit from the backtrace could be doing a better job here.
